app/controllers/checkout_controller.rb

```diff
@@ -46,26 +46,7 @@ def update
 
     fire_event('spree.checkout.update')
 
-    while @order.state != "complete"
-      if @order.state == "payment"
-        return if redirect_to_paypal_express_form_if_needed
-      end
-
-      if @order.state == "delivery"
-        @order.select_shipping_method(shipping_method_id)
-      end
-
-      next if advance_order_state(@order)
-
-      flash[:error] = if @order.errors.present?
-                        @order.errors.full_messages.to_sentence
-                      else
-                        t(:payment_processing_failed)
-                      end
-      update_failed
-      return
-    end
-    return update_failed unless @order.state == "complete" || @order.completed?
+    return unless checkout_workflow(shipping_method_id)
```

Comment:
`update_failed` is a rendering method. I personally prefer to have calls to rendering with in action method to see all possible outcomes. If you did `return update_failed unless ...` here, it would simplify the `checkout_worflow` method, reduce duplication and enable us to move that logic to a service.

Reply:
I agree with @mkllnk. Sticking to a Boolean return type in `checkout_workflow` would bring clean separation of concerns BUT this PR is amazing and I prefer to iterate instead.

Reply:
yes, I see what you mean BUT this change is related to complexity of the method. This method is already in the limit of complexity so I had to move everything to checkout_workflow.